sympy/holonomic/recurrence.py

```diff
@@ -304,7 +304,7 @@ def __eq__(self, other):
                 return False
         else:
             if self.listofpoly[0] == other:
-                for i in listofpoly[1:]:
+                for i in self.listofpoly[1:]:
```

Comment:
I've added a test, but it does not work as expected, probably because of the issue discussed in #15799 with comparing polynomials.

The idea of this part of the code is, if I understand it correctly, to compare a RecurrenceOperator with a polynomial and return True if the all but the first polynomial of the RecurrenceOperator is zero and the first polynomial is equal to the other polynomial.

Reply:
I'm looking at ways to fix #15799. Can you give me an example of how are you expecting it to work here (and how it is failing?)

Reply:
If we look at the test I added:

```
    n = symbols('n', integer=True)
    R, Sn = RecurrenceOperators(QQ.old_poly_ring(n), 'Sn')
    rr = RecurrenceOperator([n**2, 0, 0], R)
    rr2 = RecurrenceOperator([n**2, 1, n], R)
    assert not rr == rr2

    # polynomial comparison issue, see https://github.com/sympy/sympy/pull/15799
    # should work once that is solved
    # d = rr.listofpoly[0]
    # assert rr == d
```
Now, d is equal to rr.listofpoly[0] which is OK. However, then the code tests if the remaining elements are identical to the zero element of the base, which they should be, but somehow that comparison fails.

For the DifferentialOperator, I even used 'do = DifferentialOperator([x**2, R.base.zero, R.base.zero], R)' to explicitly say that those elements are zero, but to no avail.

To be honest, I do not fully grasp the polynomial rings etc, so it may be some misunderstanding on my behalf, but given the discussion in #15799 it seems related. From everything I can see (the printout), the elements look identical.

Reply:
This might be related to some of the issues that were discussed at https://github.com/sympy/sympy/pull/12088. I forgot how we ended up resolving them. I think the "proper" metaclass fix that made equality checking always work ended up being too difficult so I gave up on it. 

It's also possible that is unrelated. I'm not 100% sure if your expression is using those classes that were discussed in that PR.

Reply:
@hackalog Here comes a better explanation:

```
from sympy.holonomic.recurrence import RecurrenceOperators, RecurrenceOperator
from sympy import symbols, QQ

n = symbols('n', integer=True)
R, Sn = RecurrenceOperators(QQ.old_poly_ring(n), 'Sn')
rr = RecurrenceOperator([n**2, 0, 0], R)
```

Leads to `rr.listofpoly` is `[DMP([1, 0, 0], QQ, QQ[n]), DMP([], QQ, QQ[n]), DMP([], QQ, QQ[n])]`

In the code, equality is checked if the first polynomial of the RecurrenceOperator is equal to the compared object and the rest of the polynomials is equal to `self.parent.base.zero` which is 

```
rr.parent.base.zero
Out[5]: DMP([], QQ, QQ[n])
```

A similar thing happens in `DifferentialOperator`.

This is the equality test that doesn't work as I would expect and I guess the way it should work. (The test is written as ` i is not self.parent.base.zero` as seen below).